### ISSUE TITLE
 I'd like to add a peak finding function to scipy.signal

### ISSUE THREAD

**Issue by lagru** [scipy/scipy#8211]
Hello,

### TL;DR

I have created a [peak finding and filtering algorithm](http://nbviewer.jupyter.org/urls/gitlab.com/lagru/peakfinder-demo/raw/master/peak_finder_demo.ipynb) not unlike Matlab's [findpeaks](https://de.mathworks.com/help/signal/ref/findpeaks.html) function and would like to know if you think it worth adding to SciPy's signal module.

### Long form

It has bugged me for some time that SciPy and other scientific Python libraries lack a function for simple peak finding and filtering like Matlab's findpeaks function.

So I have come up with a solution for one of my own projects. I'm quite happy with my approach and thought it might be worth a try to find out if you would be interested in adding this in one form or another to SciPy's signal module. I have uploaded a [Jupyter notebook](http://nbviewer.jupyter.org/urls/gitlab.com/lagru/peakfinder-demo/raw/master/peak_finder_demo.ipynb) (source code [here](https://gitlab.com/lagru/peakfinder-demo/blob/master/peak_finder.py)) so you guys can see for yourself.

In my opinion the use case (simple peak finding and filtering) is currently not very well covered by SciPy and its find_peaks_cwt function. I think it would fit well into SciPy's signal module and its potential usefulness in many domains is underlined by Matlab's implementation which got extended and improved over the years.
Therefore I'd be happy to create a pull request with my work or help with something related to that.

But before I start working I'd like to know if this is actually wanted. If so, there is one more thing to discuss. I have written the algorithm as a class (as you can see in the example) because it allows a greater introspection of intermediate calculation results (e.g. the peak prominence) and a greater flexibility. However the class-based approach is rarely used througout SciPy so I'd like to know if you guys would prefer to wrap the peak finding and filtering in a single "simple" function.

I am pretty new to contributing to an open source project and I'm not claiming to be an expert in anything yet 😉 so I'd be grateful for constructive feedback in any form and color. Tips and improvements to the code itself are, of course, welcome!

### Worth noting

I haven't been able to find a discussion of this elsewhere apart from this related issue #7393. It seems like there is a general interest but not any concrete work done yet.

There are already some solutions and alternatives for peak finding out there:

* [peakutils](http://pythonhosted.org/PeakUtils/) - Python library with useful tools for peak fitting but lacks filtering functions
* [find_peaks_cwt](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.find_peaks_cwt.html) - Uses wavelet transform for peak detection, not very intuitive
* [detect_peaks](http://nbviewer.jupyter.org/github/demotu/BMC/blob/master/notebooks/DetectPeaks.ipynb) by Marcos Duarte - simple but not integrated into a library, lacks arguments to filter by shape and seems to be a little bit slower compared to my approach

(all mentioned in [this](https://blog.ytotech.com/2015/11/01/findpeaks-in-python/) blog post)

However compared to Matlab's solution these are either more complicated to use or lack basic features like peak shape filtering or integration into a common library.

**Comment by lagru** [comment]
I'm happy to hear that.

Concerning my second question: would you prefer this algorithm to be implemented as a single class or as a single / multiple functions? In my opinion there are pros and cons for both approaches:

As a class:
- Introspection and interactive data exploration is easier as intermediate results can be stored as class members
- Higher control over workflow of algorithm: order of filtering can be changed, repeated filtering with new criteria is possible
- Classmethod can act as replacement for the single-function-implementation (see `get()` method in example)
- Testing is easier (each method can be evaluated as a unit)

As one function
-  Use case is simpler but input grows more complex (all criteria need to be specified at the beginning)
- Cluttered output if more than peak locations are desired (e.g. peak statistics: prominence, width, height ...)
- If peak statistics are desired as output do I calculate everything all the time or only if used by filters?
- Matches other stuff in SciPy (I can't find many algorithms that are implemented as classes)

In my opinion the class-based approach has more advantages and the only real advantage is that it doesn't match the classic approach in SciPy. Would you agree?

**Comment by antonior92** [comment]
I agree with you.

Many SciPy functions doesn't use a class structure but there are exceptions (e.g. ``scipy.integrate.ode``). And  I like yours more "Pythonic" implementation
and I do not see any reason for not using it.  The class structure allows a nice workflow and I 
think I would prefer to use it than a matlab-like ``find_peaks`` function.

**Comment by ilayn** [comment]
That sounds nice indeed. For what its worth, you can do the same with a Python dictionary if the only thing needed is to have a container. Then all class- and static- methods become simple functions. 

I think it is an art to decide whether classes are the way to go or not. If the interface is not complicated then simple dictionaries might be preferable in the sense that 
```
mypeakfinder = Peaks()
mypeakfinder.data = myarray[3, :]
mypeakfinder.prefilters('savgol')
mypeakfinder.findpeaks()
z = mypeakfinder.peakvals
```
seems a bit strange. I am definitely not saying that this is what it would look like but I guess it is better to judge by the actual complexity instead of deciding prematurely.

Do you have a rough sketch of the general API; what the options would be, how the signature looks like and so on? It would save time and effort once that is settled instead of discussing afterwards and realizing a problem.

**Comment by antonior92** [comment]
Hi @ilayn he have posted the code [here](https://gitlab.com/lagru/peakfinder-demo/blob/master/peak_finder.py)).

**Comment by ilayn** [comment]
Oh sorry, the browser messed up the link colors. I missed them completely. I'll check

**Comment by ilayn** [comment]
That's a really nice example. Looks good to me :-) Let's wait what `scipy.signal` residents say.

**Comment by perimosocordiae** [comment]
Here are my thoughts on the API design:
 * Having multiple `filter_xx` methods isn't needed if you have a way to compute peak statistics and a generic way to filter.
 * Plotting is likely out of scope for a scipy API.
 * State management (removing peaks, restoring them, etc) is likely out of scope as well.

Basically, I'd recommend providing the bare minimum API, making sure that whatever you return is easily compatible with the standard tools. For instance, it might look like:
```python
peaks = scipy.signal.find_peaks(myarray)
peaks.centers # 1-d array
peaks.valleys  # 1-d array
heights = myarray[peaks.centers]  # or np.interp1d(myarray, peaks.centers), or just peaks.heights()
peaks.widths()
peaks.prominences()
```

The returned `peaks` object can be a custom class that knows how do compute prominences and widths and such, which are all plain numpy arrays. This gives the user all the pieces, which they can choose to assemble as needed.

**Comment by larsoner** [comment]
My 2c -- I'd prefer a function to a class. To avoid the return bloat, you can yield a `NamedTuple` or even a dict. I'd leave prefiltering and so forth to other `scipy.signal` functions.

**Comment by lagru** [comment]
@perimosocordiae 

> Plotting is likely out of scope for a scipy API.

I agree. Although it would be useful, I only included it for the demo.

> Having multiple filter_xx methods isn't needed if you have a way to compute peak statistics and a generic way to filter. [...] State management (removing peaks, restoring them, etc) is likely out of scope as well.

You have a point there and I especially like your alternative suggestion. One way to remove those filter functions and the need for state management would be to make the returned object indexable not unlike a classic `numpy.ndarray`. That way one would give the user all the tools needed. 

Example:

    >>> peaks = scipy.signal.PeakFinder(vector)  # whether find_peaks or PeakFinder shouldn't matter
    >>> peaks.indexes  # return positions of peaks in vector
    >>> peaks.prominences  # return prominences of peaks
    >>> peaks = peaks[peaks.prominences > 1.2]  # filter by prominences
    >>> peaks = peaks[peaks.height < 3]  # filter by absolute height

* Basically `PeakFinder` would accept boolean arrays (and maybe index-arrays) and return a subset of itself with the matching peaks.
* Peak stats (like prominence, width, ...) would only be computed if the user tried to access the attribute.
* This approach could store expensive intermediate results and wouldn't need to compute the same thing more than once (e.g. calculating width requires calculating prominences).
* The only case where this approach wouldn't work ist filtering by distance because that requires an iterative approach where peaks are compared with one another largest to smallest.

@larsoner 

> I'd leave prefiltering and so forth to other scipy.signal functions.

What do you mean by prefiltering? The input vector is not filtered in any way that is not required to compute all included "raw" peak positions (see function [extrema](https://gitlab.com/lagru/peakfinder-demo/blob/master/peak_finder.py#L430)). Filtering of the _peaks_ is only done after all peaks are found.

> I'd prefer a function to a class. To avoid the return bloat, you can yield a NamedTuple or even a dict. 

Why? I agree that one could return a stats object like so:

    >>> positions, stats = scipy.signal.find_peaks(vec, ...)
    >>> stats.prominences  # or stats["prominences"]

But how would we determine what to include in `stats`? Do we require the user to pass an argument telling us what to compute? Because computing everything all the time when not required would make the function unnecessary slow (computing widths, prominences can be quite expensive for a large number of peaks). One could decide to only return things like prominences if those had to be calculated for filtering any way but what if a user would want those stats without filtering by them?
If would find this behavior much more confusing and difficult to describe.

**Comment by larsoner** [comment]
> What do you mean by prefiltering? 

I was referring to the `mypeakfinder.prefilters` stuff @ilayn posted. 

> > I'd prefer a function to a class. To avoid the return bloat, you can yield a NamedTuple or even a dict.

> Why?

Your post above has a good start on some of the advantages. As you already mention, it's more consistent with how things are done in `scipy.signal`. But to that I'll add that returning standard Python / NumPy objects often simplifies downstream operations (e.g., saving with `np.savez`), the barriers for entry and access are (sometimes) lower, and so forth.

> But how would we determine what to include in stats? Do we require the user to pass an argument telling us what to compute? Because computing everything all the time when not required would make the function unnecessary slow (computing widths, prominences can be quite expensive for a large number of peaks).

Many interfaces in SciPy work via such arguments, yes. If we can cover 95% of typical use cases easily using a function with a simple interface (always return peaks, return widths if they are provided by the algorithm else None, and return prominences if asked for), that sounds better to me than implementing a class.

For example, in the context of keeping things simple yet extensible, I don't see much value in supporting peak filtering as an option directly in SciPy, at least based on the [GitLab linked code](https://gitlab.com/lagru/peakfinder-demo/blob/master/peak_finder.py). It appears to subselect found peaks using a few options for boolean masks, each computed in a handful of lines. However, in general there are many potentially useful ways to do this sort of thing (combining criteria, robust threshold choosing, etc.), and the NumPy code to do the selection step appears simple. It would be better to show users how to do this sort of filtering in examples (e.g., you now have the peaks, widths, and prominences, here is how you can subselect/filter peaks using these quantities in a few lines!) rather than support these with SciPy wrappers.

**Comment by lagru** [comment]
Ah, so basically the new function would be responsible for finding peak positions and potentially computing additional stats for those found peaks:

```python
>>> peaks, prominences, widths = scipy.signal.find_peaks(vec, proms=True, widths_at="halfprom")
```
The problem here is that for noisy signals computing prominences / widths would be quite expensive without limiting the peak number by filtering with cheaper criteria first.

Or another more verbose / extensible approach which could avoid the problem above:

```python
minima, maxima = scipy.signal.extrema(vec)
maxima = scipy.signal.minimal_peak_distance(vec, maxima, min_=200)
prominences = scipy.signal.peak_prominence(vec, maxima)
widths = scipy.signal.peak_width(vec, maxima, at="halfprom")  
# prominences computed if not provided
```

I can see how would simplify the the implementation and still provide the user with the tools for calculating peak prominence, etc.

However that would make the main use case peak detection and filtering (only a subset of all local maxima are wanted peaks) more complicated and we'd have to provide extensive examples to even showcase how to filter peaks. And we would still have to provide a function for filtering by minimal peak distance as we can't expect every user to implement this himself.  I'm not convinced this sacrifice would be worth it "just" to avoid using a class.

I think you guys have way more experience in judging these things and you are the ones that have to accept and review the potential pull request. All suggested APIs would be "easy" to implement with the code that is provided in the demo.
So how do you usually go about deciding this? And what is the opinion of the majority?

**Comment by antonior92** [comment]
> One way to remove those filter functions and the need for state management would be to make the returned object indexable not unlike a classic numpy.ndarray. That way one would give the user all the tools needed.
> - Basically PeakFinder would accept boolean arrays (and maybe index-arrays) and return a subset of itself with the matching peaks.
> - Peak stats (like prominence, width, ...) would only be computed if the user tried to access the attribute.
> - This approach could store expensive intermediate results and wouldn't need to compute the same thing more than once (e.g. calculating width requires calculating prominences).
> - The only case where this approach wouldn't work ist filtering by distance because that requires an iterative approach where peaks are compared with one another largest to smallest.

@lagru this seems like an elegant implementation for me. And it seems a much better approach than the two others you proposed latter trying to use only numpy arrays.

My general feeling is that having a implementation without using standard python objects may be reasonable if it result in a much easier usage and cleaner implementation.

**Comment by larsoner** [comment]
> Or another more verbose / extensible approach which could avoid the problem above: [snip: function-based version]

This looks reasonable start to me. I don't see a need for `minimal_peak_distance`, though.

> However that would make the main use case peak detection and filtering (only a subset of all local maxima are wanted peaks) more complicated and we'd have to provide extensive examples to even showcase how to filter peaks. And we would still have to provide a function for filtering by minimal peak distance as we can't expect every user to implement this himself. 

I disagree here. I think that peak filtering is something we can and should expect users to be able to do.
We can say something like:
```
To filter peaks based on some minimum distance criteria, you can do for example::

    >>> maxima = maxima[np.concatenate(([True, np.diff(maxima) >= 200]))`
```
This doesn't seem too onerous. Again users can want to index and subselect in all sort of ways. I'd rather teach them the general principle of how to do it -- especially if it only takes a handful (or one!) line -- than try to implement every option in SciPy.

> My general feeling is that having a implementation without using standard python objects may be reasonable if it result in a much easier usage and cleaner implementation.

I agree with this sentiment, but I don't think it applies here. I think that the usage is arguably simpler in the function case because it operates on standard objects (that incur no "how do I use this thing" overhead).

In the class-based indexing proposal, for example, having a class that says "you can index me like a NumPy array!" but isn't actually a NumPy array will cause more problems than it solves. You will likely not support every way one expects to operate on NumPy arrays (slicting, boolean indexing, expansion of dimensions, resizing, recasting to types, etc.) so now you have a partial implementation with some unexpected surprises during use. Users find out what works and what doesn't by trial and error. You have lost the ability to save it with `np.save` or `h5py` or any other standard utility.

> So how do you usually go about deciding this? And what is the opinion of the majority?

Ideally people will become convinced one way or the other and we can reach a consensus.

**Comment by lagru** [comment]
> We can say something like: [code snippet: `>>> maxima = ...`]

If you look at the [implementation of distance filtering](https://gitlab.com/lagru/peakfinder-demo/blob/master/peak_finder.py#L151) you'll notice that in contrast to your one line version the peak elimination order is depending on their amplitude. If two peaks are to close to another the lower one is removed. That's actually one of the more useful filter options (e.g. noisy biosignals like an EKG where  a certain peak periodicity is expected). And while one doesn't need to invoke black magic to implement that I didn't find it trivial either.

>[...] now you have a partial implementation with some unexpected surprises [...] "you can index me like a NumPy array!" but isn't actually a NumPy array!" [...]

If we would use an extra method perhaps named `.filter(index_arg)` instead of Pythons `__getitem__` it would be clear that the returned class is not in any way an array. That way we would clearly differentiate the class from a numpy array while enabling the user and keeping the extensibility.

> You have lost the ability to save it with np.save or h5py or any other standard utility.

Why? Sure we can't save the direct class itself with those functions but we can save all relevant results (peak positions, prominence, ...). I would see the class more as a fancy function with state management and don't really see the use case as common where a user wants to save a `PeakFinder` object to latter reload and resume usage of that class. 

----

Maybe an API that all could agree on would look like this:

* `minima, maxima = signal.extrema(vec)` - returns all (unfiltered) positions of extreme points in `vec`
* `proms = signal.peak_prominences(peaks, vec)` - computes prominences of provided local maxima
* `widths = signal.peak_widths(peaks, vec, at="halfprom")` - calculates with at a certain height relative to peak height or prominence (no one-liner)
* `peaks = signal.filter_by_peak_distance(peaks, dmin=200)` - as described above, I don't like the name but can't think of a better one

These functions would be self contained and part of the official API. Compared to the above functions threshold and height filtering are trivial so we could expect the user to do that himself. 

Now the reason for this entire issue was to provide the user with a convenient solution equivalent to Matlab's findpeaks function and other implemenations. So I would propose that we wrap the above functions together with either another function: 
`peaks, stats = signal.find_peaks(vec, ..., stats=True)`. 
If `stats=True`, intermediate results that had to be calculated anyway for filtering (prominences, widths, ...) would be returned as well.

----

This would:
* enable peak finding & filtering in one line in Python making it convenient
* provide the user with all the moving parts for data exploration and more complex use cases that aren't wouldn't be covered by `find_peaks`.
* keep unit tests relatively simple because no state management needs to be covered.
* keep the API consistent.

Also, `find_peaks` could be implemented as a class  as detailed in the above discussion which I think would be the more pythonic and convenient solution. But considering the context I think the proposed solution above fits the scope of SciPy better.

**Comment by antonior92** [comment]
>  having a class that says "you can index me like a NumPy array!" but isn't actually a NumPy array will cause more problems than it solves. You will likely not support every way one expects to operate on NumPy arrays (slicting, boolean indexing, expansion of dimensions, resizing, recasting to types, etc.) so now you have a partial implementation with some unexpected surprises during use. 

I agree with you there @larsoner . I have not though about this.

-----------------------------------

> If you look at the implementation of distance filtering you'll notice that in contrast to your one line version the peak elimination order is depending on their amplitude. If two peaks are to close to another the lower one is removed. That's actually one of the more useful filter options (e.g. noisy biosignals like an EKG where a certain peak periodicity is expected). And while one doesn't need to invoke black magic to implement that I didn't find it trivial either.

In this case I think @lagru have a good point. I would like to have this already implemented. It is not so trivial to implement and, besides that, it is needed often when dealing with signal processing of noisy signals.

 -------------------

I kind of disagree with the others that you should let the user implement filtering options using numpy operations. This options are not so trivial to implement and since they are used frequently it would be nice to have them implemented in SciPy. Matlab does gives these options to the user and I think it does a great job with its ``findpeaks`` function.

Hence I think it would be good to include more options to ``find_peaks`` function you propose. Something like you have implemented in your current ``__init__`` function:
```
peaks, stats = signal.find_peaks(vec,  filter_by_weight=(min, max), filter_by_threshold=tmin, filter_by_distance=dmin, filter_by_proeminence=(pmin, pmax, wlen), filter_by_width=(wmin, wmax, wheight), return_width=True, return_proeminence=False)
```

I do not see any harm in include filtering options  like ``filter_by_weight`` or ``filter_by_distance``. If the user do not want to use them he can just left this options untouched. It will not warm his ability of using custom filters on the return data.

**Comment by lagru** [comment]
It sounds like you would agree with the following **API** draft:
- A basic function to find local maxima (/minima) in a vector. It seems like something like this is already provided with [argrelextrema](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.argrelextrema.html#scipy.signal.argrelextrema) which I could use for my first pull request. Later one could look at my own implementation of `extrema` and evaluate which approach is faster / more efficient.
- Coherent functions to compute peak properties, meaning:
  - peak prominence
  - peak width
  - peak threshold, quite simple to do oneself but may still be worth it
- A function that filters peaks by their distance (priority by peak height). I can't decide whether this function should be public or underscored. In any case users would have access to with the next function...
- A `find_peaks` function that wraps peak finding and filtering in one convenient function call.

This has the following **Advantages**:
-  All basic use cases would be covered with `find_peaks` which would provide an interface coherent with the rest of SciPy and would provide minimal surprises for users coming from Matlab. 
- For more complicated use cases and workflows the underlying functions would be easily accessible.

@larsoner would you agree? 

I encourage any additional feedback and thoughts before I start working on a pull request!

**Comment by perimosocordiae** [comment]
Sounds like a good plan to me. Feel free to open a PR as soon as you're ready for feedback.

**Comment by larsoner** [comment]
Sorry for the long delay, was away for a bit, yes this sounds good to me
